### DISCUSSION TITLE
How to plot thousands of separate lines in 3D

### DISCUSSION THREAD

**Discussion by Breviel** [pyvista/pyvista#2387] [Q&A] [answered]
Hey,

I have 2 arrays of 19000 XYZ coordinates, one for start point of line and other for end point of line.
I would like to plot those lines in 3D and them to be drawn at once. It's possible to add each line separately with for loop with
pyvista.Line(start, end), but that is not an efficient method.

I would like it to look similar to this. The lines themselves are not connected.

![lines](https://user-images.githubusercontent.com/51784217/160382137-a771b0c0-8afe-4e89-a4c0-b5ff330930b5.jpg)

Appreciate all the help I can get to solve this :)

**Comment by akaszynski** [comment, marked as answer]
Hi @Breviel,

The best approach would be to create a [pyvista.PolyData](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyData.html) and initialize from vertices and lines.

```py
import numpy as np
import pyvista
vertices = np.array([[0, 0, 0], [1, 0, 0], [1, 0.5, 0], [0, 0.5, 0]])
lines = np.hstack([[2, 0, 1], [2, 1, 2]])
mesh = pyvista.PolyData(vertices, lines=lines)
```

This defines two lines `[2, 0, 1]` and `[2, 1, 2]` where the first number denotes the number of points in a line segment and the next two numbers are the indices of those points.

**Reply by Breviel** [reply]
I got it to work with pyvista.line_segments_from_points() function :). Generates nicely all the same lines as in the example above.
![line segments](https://user-images.githubusercontent.com/51784217/160703942-34416c55-9fa6-4c26-ad2a-c346d789fc6e.jpg)

**Reply by akaszynski** [reply]
Nice work!

**Comment by MatthewFlamm** [comment]
If you want to eventually use another shape type like a cylinder or ribbon, you could also consider using the glyph filter: https://docs.pyvista.org/api/core/_autosummary/pyvista.DataSet.glyph.html

**Comment by Breviel** [comment]
Thank you for the answers @akaszynski and @MatthewFlamm , I will give both a try. 

Cheers
